Added CosmicLivetime data product; Clean-up of CORSIKA code

soleti opened the pull request:
The live-time of a cosmic sample can be calculated from 5 parameters, according to the cosmic-ray primary nucleon intensity formula (eq. 29.2 of http://pdg.lbl.gov/2018/reviews/rpp2018-rev-cosmic-rays.pdf). These parameters are:
- the number of primary nucleons
- the sampled area (in m2)
- the minimum energy (in GeV)
- the maximum energy (in GeV)
- the flux constant (1.8e4 by default).

This pull request includes a new data product with these 5 parameters, plus the calculated live-time. The data product is stored for each subrun.

oksuzian commented on Sources/inc/CosmicCORSIKA.hh:
```diff
@@ -47,22 +46,24 @@ struct Config
-  fhicl::Atom<float> targetBoxYmin{Name("targetBoxYmin"), Comment("Extension of the generation plane"), -5000};
-  fhicl::Atom<float> targetBoxYmax{Name("targetBoxYmax"), Comment("Extension of the generation plane"), 5000};
-  fhicl::Atom<float> targetBoxZmin{Name("targetBoxZmin"), Comment("Extension of the generation plane"), -5000};
-  fhicl::Atom<float> targetBoxZmax{Name("targetBoxZmax"), Comment("Extension of the generation plane"), 5000};
```
Why don't you use the same definition of target Box as in defaultCRYconfig?

soleti replied:
These are the default values, but they definition in the FCL file is the same as defaultCRYconfig. However, to avoid confusion, I changed also the values here. Also, the comment is wrong and has been fixed.

oksuzian commented on EventGenerator/test/corsikaTest.fcl:
```diff
@@ -17,14 +17,16 @@ targetParams: {
 source: {
     module_type: FromCorsikaBinary
-    fileNames: ["/mu2e/app/users/srsoleti/corsika-77100/run/DAT110001"]
+    fileNames: ["/mu2e/data/users/srsoleti/DAT110001"]
```
Why don't you move it to persistent?

soleti replied:
Moved to persistent.

oksuzian commented on MCDataProducts/inc/CosmicLivetime.hh:
```diff
@@ -0,0 +1,78 @@
+        const float EiToOneMinusGamma = pow(_lowE, 1 + eslope);
+        const float EfToOneMinusGamma = pow(_highE, 1 + eslope);
+        // http://pdg.lbl.gov/2018/reviews/rpp2018-rev-cosmic-rays.pdf eq. 29.2
+        _livetime = _primaries / (M_PI * _area * _fluxConstant * (EfToOneMinusGamma - EiToOneMinusGamma) / (1. + eslope));
```
The integral of eq 29.2 looks right to me.
However, have you already run the cross-checks with CRY and Daya Bay?

soleti replied:
I've normalized the events to the same cosmic live-time for CRY and CORSIKA in the plots shown in https://mu2e-docdb.fnal.gov/cgi-bin/private/RetrieveFile?docid=29724&filename=mu2emeeting20191106.pdf&version=1.

oksuzian replied:
I remember seeing your slides.
I have not realized the comparisons are based on that formula.
Looks good.

gaponenko commented:
Hi Roberto,

The Mu2e policy is to have all "physics" parameters in fcl without 
defaults in C++.  Only "debug" parameters like verbosityLevel are allowed 
to have C++ defaults.

Andrei

On Tue, 12 Nov 2019, Stefano Roberto Soleti wrote:

> soleti commented on this pull request.
> 
> 
> 
> >    fhicl::Atom<float> fluxConstant{Name("fluxConstant"), Comment("Flux constant"), 1.8e4};
> -  fhicl::Atom<float> targetBoxXmin{Name("targetBoxXmin"), Comment("Extension of the generation plane"), -5000};
> -  fhicl::Atom<float> targetBoxXmax{Name("targetBoxXmax"), Comment("Extension of the generation plane"), 5000};
> -  fhicl::Atom<float> targetBoxYmin{Name("targetBoxYmin"), Comment("Extension of the generation plane"), -5000};
> -  fhicl::Atom<float> targetBoxYmax{Name("targetBoxYmax"), Comment("Extension of the generation plane"), 5000};
> -  fhicl::Atom<float> targetBoxZmin{Name("targetBoxZmin"), Comment("Extension of the generation plane"), -5000};
> -  fhicl::Atom<float> targetBoxZmax{Name("targetBoxZmax"), Comment("Extension of the generation plane"), 5000};
> 
> These are the default values, but they definition in the FCL file is the same as defaultCRYconfig. However, to avoid confusion, I changed also the values here. Also, the comment is wrong and has been fixed.
> 
> -- 
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly or view it on GitHub:
> https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_Mu2e_Offline_pull_51-23discussion-5Fr345471324&d=DwICaQ&c=gRgGjJ3BkIsb5y6s49QqsA&r=O47fc5vzDTR2V_gla4Ub0Q&m=_hS37U3TjxYM31Tv3p1IubPotJdqtl3tXSf1g5n33xc&s=mVWnHDCa9cOJunwQKt5-tUyQxwc-WvQYV_Q6ZgMjP_4&e=

brownd1978 commented:
On Wed, Nov 13, 2019 at 9:50 AM gaponenko <notifications@github.com> wrote:

> Hi Roberto,
>
> The Mu2e policy is to have all "physics" parameters in fcl without
> defaults in C++. Only "debug" parameters like verbosityLevel are allowed
> to have C++ defaults.
>

parameters whose values only change for special purposes are also allowed
code defaults.  That may be a generalization of 'debug'.
Dave


>
> Andrei
>
> On Tue, 12 Nov 2019, Stefano Roberto Soleti wrote:
>
> > soleti commented on this pull request.
> >
> >
> >
> > > fhicl::Atom<float> fluxConstant{Name("fluxConstant"), Comment("Flux
> constant"), 1.8e4};
> > - fhicl::Atom<float> targetBoxXmin{Name("targetBoxXmin"),
> Comment("Extension of the generation plane"), -5000};
> > - fhicl::Atom<float> targetBoxXmax{Name("targetBoxXmax"),
> Comment("Extension of the generation plane"), 5000};
> > - fhicl::Atom<float> targetBoxYmin{Name("targetBoxYmin"),
> Comment("Extension of the generation plane"), -5000};
> > - fhicl::Atom<float> targetBoxYmax{Name("targetBoxYmax"),
> Comment("Extension of the generation plane"), 5000};
> > - fhicl::Atom<float> targetBoxZmin{Name("targetBoxZmin"),
> Comment("Extension of the generation plane"), -5000};
> > - fhicl::Atom<float> targetBoxZmax{Name("targetBoxZmax"),
> Comment("Extension of the generation plane"), 5000};
> >
> > These are the default values, but they definition in the FCL file is the
> same as defaultCRYconfig. However, to avoid confusion, I changed also the
> values here. Also, the comment is wrong and has been fixed.
> >
> > --
> > You are receiving this because you are subscribed to this thread.
> > Reply to this email directly or view it on GitHub:
> >
> https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_Mu2e_Offline_pull_51-23discussion-5Fr345471324&d=DwICaQ&c=gRgGjJ3BkIsb5y6s49QqsA&r=O47fc5vzDTR2V_gla4Ub0Q&m=_hS37U3TjxYM31Tv3p1IubPotJdqtl3tXSf1g5n33xc&s=mVWnHDCa9cOJunwQKt5-tUyQxwc-WvQYV_Q6ZgMjP_4&e=
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/Mu2e/Offline/pull/51?email_source=notifications&email_token=ABAH57Y664RTPRDCPXP3FXLQTQ4HJA5CNFSM4JL4IC62YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOED7A5RY#issuecomment-553520839>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABAH572ANIW2I5OOB6E7KA3QTQ4HJANCNFSM4JL4IC6Q>
> .
>


-- 
David Nathan Brown            Dave_Brown@lbl.gov
Office Phone (510) 486-7261  Fax  495-2957
Lawrence Berkeley National Lab
MS 50R5008 (50-6026C)  Berkeley, CA 94720

soleti commented:
I removed the default values for the geometric parameters and I added two features:
- the possibility to stop the backtracking if two or more particles trajectories have a point of closest approach smaller than a fixed value. This is turned off at the moment.
- A random shift on the xz axes for the particles in the same shower.

soleti commented:
Three main changes:
- Added the DCA of the tracks to CORSIKAGenPlots
- Added the possibility to resample the same CORSIKA binary
- Fixed the cosmic livetime issue: now the number of primaries is stored per event and the cosmic livetime calculation is done at producer level at the end of the Subrun.
